test/framework/machinehealthcheck_helpers.go

```diff
@@ -0,0 +1,169 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type ApplyMachineHealthCheckAndWaitInput struct {
+	ClusterProxy              ClusterProxy
+	Cluster                   *clusterv1.Cluster
+	ControlPlane              *controlplanev1.KubeadmControlPlane
+	MachineDeployments        []*clusterv1.MachineDeployment
+	WaitForMachineDeployments []interface{}
+}
+
+func ApplyMachineHealthCheckAndWait(ctx context.Context, input ApplyMachineHealthCheckAndWaitInput) {
+	Expect(ctx).NotTo(BeNil(), "ctx is required for ApplyMachineHealthCheckAndWait")
+	Expect(input.ClusterProxy).ToNot(BeNil(), "Invalid argument. input.ClusterProxy can't be nil when calling ApplyMachineHealthCheckAndWait")
+	Expect(input.Cluster).ToNot(BeNil(), "Invalid argument. input.Cluster can't be nil when calling ApplyMachineHealthCheckAndWait")
+	Expect(input.MachineDeployments).ToNot(BeEmpty(), "Invalid argument. input.MachineDeployments can't be empty when calling ApplyMachineHealthCheckAndWait")
+
+	fmt.Fprintf(GinkgoWriter, "Creating MachineHealthCheck instance\n")
+	mhc := GenerateMachineHealthCheck(input.Cluster)
+	Expect(input.ClusterProxy.GetClient().Create(ctx, mhc)).ShouldNot(HaveOccurred())
+
+	fmt.Fprintf(GinkgoWriter, "Patching an unhealthy condition to nodes and waiting for remediation\n")
+	WaitForMachineHealthCheckToDetectUnhealthyNodeCondition(ctx, WaitForMachineHealthCheckToDetectUnhealthyNodeConditionInput{
+		ClusterProxy:           input.ClusterProxy,
+		Cluster:                input.Cluster,
+		ControlPlane:           input.ControlPlane,
+		MachineDeployments:     input.MachineDeployments,
+		MachineHealthCheckName: mhc.Name,
+	}, input.WaitForMachineDeployments...)
+}
+
+func GenerateMachineHealthCheck(cluster *clusterv1.Cluster) *clusterv1.MachineHealthCheck {
+	mhc := &clusterv1.MachineHealthCheck{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("mhc-%s%s", cluster.Name, util.RandomString(6)),
+			Namespace: cluster.Namespace,
+		},
+		Spec: clusterv1.MachineHealthCheckSpec{
```

Comment:
I was expecting to have a machine selector here
(I don't get how this match with input.MachineDeployments)
Also, are we expecting all the machines to fail?

Reply:
Without the selector, it matches all machines. But we only add failing condition to worker machines, so only they are expected to fail.
Changing MHC logic. Will create another PR for MCH test.

Reply:
Probably best to set a selector just in case these tests ever end up being run in parallel with other tests that may also cause unhealthy nodes

Reply:
+1 to use selectors.
I was expecting us forcing to fail a single machine, and watch this machine being replaced (by MD or KCP in future). If we are forcing all the machines to fail I think it is harder to follow what is going on...

Reply:
Okay, I changed it to use selectors and separated MHC tests from this PR. After this one gets merged, will create that one.